lib/auth/keystore/gcp_kms.go

```diff
@@ -0,0 +1,454 @@
+// configured with the same keyring. This is a divergence from the PKCS#11
+// keystore where different auth servers will always create their own keys even
+// if configured to use the same HSM
+func (g *gcpKMSKeyStore) canSignWithKey(raw []byte, keyType types.PrivateKeyType) (bool, error) {
+	if keyType != types.PrivateKeyType_GCP_KMS {
+		return false, nil
+	}
+	keyID, err := parseGCPKMSKeyID(raw)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	if !strings.HasPrefix(keyID.keyVersionName, g.keyRing) {
+		return false, nil
+	}
+	return true, nil
+}
+
+// DeleteUnusedKeys deletes all keys from KMS if they are:
+// 1. Labeled by this server (matching HostUUID) when they were created
+// 2. Not included in the argument usedKeys
+func (g *gcpKMSKeyStore) DeleteUnusedKeys(ctx context.Context, usedKeys [][]byte) error {
+	usedKmsKeyVersions := make(map[string]struct{})
+	for _, usedKey := range usedKeys {
+		keyID, err := parseGCPKMSKeyID(usedKey)
+		if err != nil {
+			// could be a different type of key
+			continue
+		}
+		usedKmsKeyVersions[keyID.keyVersionName] = struct{}{}
+	}
+
+	var unusedKeyIDs []gcpKMSKeyID
+
+	listKeyRequest := &kmspb.ListCryptoKeysRequest{
+		Parent: g.keyRing,
+		Filter: fmt.Sprintf("labels.%s=%s", hostLabel, g.hostUUID),
+	}
+	iter := g.kmsClient.ListCryptoKeys(ctx, listKeyRequest)
+	key, err := iter.Next()
+	for err == nil {
+		keyVersionName := key.Name + keyVersionSuffix
+		if _, used := usedKmsKeyVersions[keyVersionName]; !used {
+			unusedKeyIDs = append(unusedKeyIDs, gcpKMSKeyID{
+				keyVersionName: keyVersionName,
+			})
+		}
+		key, err = iter.Next()
+	}
+	if err != nil && !errors.Is(err, iterator.Done) {
+		return trace.Wrap(err)
+	}
+
+	for _, unusedKey := range unusedKeyIDs {
+		g.log.WithFields(logrus.Fields{"key_version": unusedKey.keyVersionName}).Info("deleting unused GCP KMS key created by this server")
+		err := g.deleteKey(ctx, unusedKey.marshal())
+		// Ignore errors where we can't destroy because the state is already
+		// DESTROYED or DESTROY_SCHEDULED
+		if err != nil && !strings.Contains(err.Error(), "has value DESTROY") {
+			g.log.WithFields(logrus.Fields{"key_version": unusedKey.keyVersionName}).WithError(err).Warn("error deleting unused GCP KMS key")
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// kmsSigner implements the crypto.Signer interface
+type kmsSigner struct {
+	kmsClient kmsClientI
+	keyID     gcpKMSKeyID
+	public    crypto.PublicKey
+	clock     clockwork.Clock
+}
+
+func (g *gcpKMSKeyStore) newKmsSigner(ctx context.Context, keyID gcpKMSKeyID) (*kmsSigner, error) {
+	req := &kmspb.GetPublicKeyRequest{
+		Name: keyID.keyVersionName,
+	}
+	// Retry fetching the public key while the key is pending creation.
+	resp, err := retryWhilePending(ctx, g.clock, g.kmsClient.GetPublicKey, req)
+	if err != nil {
+		return nil, trace.Wrap(err, "unexpected error fetching public key")
+	}
+
+	block, _ := pem.Decode([]byte(resp.Pem))
+	if block == nil {
```

Comment:
```suggestion
	if block == nil || block.Type != "PUBLIC KEY or whatever" {
```

Reply:
I would expect `ParsePKIXPublicKey` to return the error in that case, and I don't expect GCP to give us something other than a public key, any particular concern here? I'm just checking for `nil` to avoid the panic.

Reply:
Still a bit concerned about potentially parsing some random ASN.1 as a public key, but I guess that if it walks and quacks like a duck we might as well roast it.

Reply:
Not exactly random, it's google's response to the GetPublicKey endpoint. I think I'd rather keep it like this in case they change the type from "PUBLIC KEY" to "RSA PUBLIC KEY" or something, not that anyone (teleport) has done something like that before